src/freenet/client/filter/ContentFilter.java

```diff
@@ -343,16 +344,8 @@ public static FilterStatus filter(InputStream input, OutputStream output, String
 			if(handler.readFilter != null) {
 				if(handler.takesACharset && ((charset == null) || (charset.isEmpty()))) {
 					int bufferSize = handler.charsetExtractor.getCharsetBufferSize();
-					input.mark(bufferSize);
 					byte[] charsetBuffer = new byte[bufferSize];
-					int bytesRead = 0, offset = 0, toread=0;
-					while(true) {
-						toread = bufferSize - offset;
-						bytesRead = input.read(charsetBuffer, offset, toread);
-						if(bytesRead == -1 || toread == 0) break;
-						offset += bytesRead;
-					}
-					input.reset();
+					int offset = readIntoBuffer(input, bufferSize, charsetBuffer);
 					charset = detectCharset(charsetBuffer, offset, handler, maybeCharset);
```

Comment:
I believe the correct solution to this problem is moving this block of code right before the `if(handler.readFilter != null)` check: text/plain does not have a `readFilter`, but does `takesACharset` so this would run the `detectCharset` appropriately.

Few things to consider:
* `handler.charsetExtractor.getCharsetBufferSize()` will NPE, so we need to choose the bufferSize to the max BOM length (5?) when `handler.charsetExtractor` is absent.
* Alternatively a dummy CharsetExtractor can be used that always fails to detect a charset, so the BOM one is used automagically.
* this will return `UTF-8` rather than `utf-8` so the related test would need some adjustment.